### PR TITLE
fix(sdk): Fix long resource name

### LINKED ISSUE #518: Resource names are too long
/kind bug

**What steps did you take and what happened:**
Create any pipeline with a long name (over 63 characters after sanitisation) which uses `dsl.ParallelFor`:

```python
from typing import List

from kfp import dsl
from kfp_tekton.compiler import TektonCompiler as Compiler

class PrintOp(dsl.ContainerOp):
    def __init__(self, name, msg):
        super(PrintOp, self).__init__(
            name=name,
            image='alpine:3.6',
            command=['echo', msg])

@dsl.pipeline(name="Some very long name with lots of words in it. " +
                   "It should be over 63 chars long in order to observe the problem.")
def main_fn(arr: List[str] = ["a", "b", "c"]):
    with dsl.ParallelFor(arr) as it:
        PrintOp('print', it)

if __name__ == '__main__':
    Compiler().compile(main_fn, __file__+'.yaml')
```

Compile it.

Created files:
* `long_names_with_resources.py.yaml`
* `long_names_with_resources.py_pipelineloop_cr1.yaml`

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: some-very-long-name-with-lots-of-words-in-it-it-should-be-o
  annotations:
    tekton.dev/output_artifacts: '{}'
    tekton.dev/input_artifacts: '{}'
    tekton.dev/artifact_bucket: mlpipeline
    tekton.dev/artifact_endpoint: minio-service.kubeflow:9000
    tekton.dev/artifact_endpoint_scheme: http://
    tekton.dev/artifact_items: '{"print": []}'
    sidecar.istio.io/inject: "false"
    pipelines.kubeflow.org/pipeline_spec: '{"inputs": [{"default": "[\"a\", \"b\",
      \"c\"]", "name": "arr", "optional": true, "type": "typing.List[str]"}], "name":
      "Some very long name with lots of words in it. It should be over 63 chars long
      in order to observe the problem."}'
spec:
  params:
  - {name: arr, value: '["a", "b", "c"]'}
  pipelineSpec:
    params:
    - {name: arr, default: '["a", "b", "c"]'}
    tasks:
    - name: some-very-long-name-with-lots-of-words-in-it-it-should-be-over-799fe-for-loop-1
      taskRef: {apiVersion: custom.tekton.dev/v1alpha1, kind: PipelineLoop, name: some-very-long-name-with-lots-of-words-in-it-it-should-be-over-799fe-for-loop-1}
      params:
      - {name: arr-loop-item, value: $(params.arr)}
  timeout: 0s
```
```
apiVersion: custom.tekton.dev/v1alpha1
kind: PipelineLoop
metadata: {name: some-very-long-name-with-lots-of-words-in-it-it-should-be-over-799fe-for-loop-1}
spec:
  pipelineSpec:
    params:
    - {name: arr-loop-item, type: string}
    tasks:
    - name: print
      params:
      - {name: arr-loop-item, value: $(params.arr-loop-item)}
      taskSpec:
        steps:
        - name: main
          command: [echo, $(inputs.params.arr-loop-item)]
          image: alpine:3.6
        params:
        - {name: arr-loop-item, type: string}
      timeout: 0s
  iterateParam: arr-loop-item
```

Note that PipelineRun's name got shortened:
```python
>>> len("some-very-long-name-with-lots-of-words-in-it-it-should-be-o") 
59
```
But the loop resource's way above the limit:
```python
>>> len("some-very-long-name-with-lots-of-words-in-it-it-should-be-over-799fe-for-loop-1")
79
```

Names that long lead to a runtime error in Kubernetes:
```
cannot create resource with name =
  some-very-long-name-with-lots-of-words-in-it-it-should-be-over-799fe-for-loop-1
gvk=custom.tekton.dev/v1alpha1, Kind=PipelineLoop: admission webhook
\"validation.webhook.tektonextensions.custom.tekton.dev\" denied the request: validation failed:
Invalid resource name: length must be no more than 63 characters: metadata.name
```

**What did you expect to happen:**

Resource names should be shortened too. It may require to shorten the prefix, which is got from PipelineRun name, for all resources to fit in the limit.

**Additional information:**
Looks like resource names are not shortened at all right now.

It may be a good idea to shorten the names before the templates are generated (rather than as post-processing), by iterating over all the DSL structure to check which resource will have the longest name and shortening the `PipelineRun`'s name accordingly.

Considering how the longest the suffix can get is `-(a fixed number of random digits)-(fixed name of group type)-(some number)`, and the last number is probably no longer than 2-3 digits, the total length of the suffix should be no longer than 17-20 characters. That leaves 53 characters.

**Environment:**

* Python Version (use `python --version`):
* SDK Version: 
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by Tomcli**
**Which issue is resolved by this Pull Request:** 
Resolves #518 

**Description of your changes:**
Fix resource names that will go beyond the k8s 63 chars limit. On both local and KFP runs.

**Environment tested:**

* Python Version (use `python --version`): 3.8
* Tekton Version (use `tkn version`): 0.21.0
* Kubernetes Version (use `kubectl version`): 1.18
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by Tomcli** [comment]
FYI @ckadner

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 1]
```diff
@@ -1001,6 +1002,7 @@ def _create_and_write_workflow(self,
         params_list,
         pipeline_conf)
     # Separate loop workflow from the main workflow
+    print(self.loops_pipeline)
```
Comment (Udiknedormin) [inline sdk/python/kfp_tekton/compiler/compiler.py:1005]: Why does kfp-tekton produce that to standard output?

Comment (Tomcli) [reply sdk/python/kfp_tekton/compiler/compiler.py:1005]: sorry it was a mistake, I removed it at https://github.com/kubeflow/kfp-tekton/pull/524

**Pull request merged**